custom-recipes/send-mails-from-contacts-dataset/recipe.py

```diff
@@ -31,88 +103,84 @@
 subject_value = config.get('subject_value', None)
 use_subject_value = config.get('use_subject_value', False)
 
+use_body_value = config.get('use_body_value', False)
+
 body_column = config.get('body_column', None)
 body_value = config.get('body_value', None)
-body_encoding = config.get('body_encoding', 'us-ascii')
-use_body_value = config.get('use_body_value', False)
 
-smtp_host = config.get('smtp_host', None)
-smtp_port = int(config.get('smtp_port', "25"))
+# For legacy configs, assume it is text if not defined
+body_format = config.get('body_format', 'text')
+# For sending a body from a column value we also assume it is plain text - that is the legacy behavour
+use_html_body_value = use_body_value and (body_format == 'html')
 
-smtp_use_tls = config.get('smtp_use_tls', False)
-smtp_use_auth = config.get('smtp_use_auth', False)
-smtp_user = config.get('smtp_user', None)
-smtp_pass = config.get('smtp_pass', None)
+html_body_value = config.get('html_body_value', None)
+mail_channel = config.get('mail_channel', None)
+channel_has_sender = does_channel_have_sender(mail_channel)
 
 attachment_type = config.get('attachment_type', "csv")
 
-output_schema = list(people.read_schema())
-output_schema.append({'name': 'sendmail_status', 'type': 'string'})
-output_schema.append({'name': 'sendmail_error', 'type': 'string'})
-output.write_schema(output_schema)
+# Validation part 1 - Check some kind of value/column exists for body, subject, sender and recipient
 
-if not body_column and not body_value:
+has_body_column = body_column and not use_body_value
+has_plain_body_value = use_body_value and body_value and not use_html_body_value
+has_html_body_value = use_body_value and use_html_body_value and html_body_value
+if not (has_body_column or has_plain_body_value or has_html_body_value):
```

Comment:
I see what you mean. When I was writing I thought it was clear as there are 3 UI states we need to check for validity and this made that very explicit. But now I realise that it would be clearer to do in if statements as follows - this would reflect the choices present in the UI and is easier to follow I think - do you agree?

```
# Check a body is provided for the options selected
is_body_present = False
if use_body_value:
    if use_html_body_value:
        is_body_present = bool(html_body_value)
    else:
        is_body_present = bool(body_value)
else:
    is_body_present = bool(body_column)

if not is_body_present:
    raise AttributeError("No body column nor body value specified")
```

Reply:
Done [here](https://github.com/dataiku/dss-plugin-sendmail/pull/8/commits/e4fe6e3751e9e11b47a4d603404955e57311514e)

(Definitely clearer this way)